File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        int len = s.size();
        vector<bool> present(26, false);
        vector<int> lastOccurrence(26, 0);
        for (int i = 0; i < len; i++) {
            char ch = s[i];
            lastOccurrence[ch - 'a'] = i;
        }

        stack<char> st;
        for (int i = 0; i < len; i++) {
            char ch = s[i];

            if (st.empty()) {
                st.push(ch);
                present[ch - 'a'] = true;
            } else {
                while (!st.empty() && ch < st.top() && lastOccurrence[st.top() - 'a'] > i && !present[ch - 'a']) {
                    char top = st.top();
                    st.pop();
                    present[top - 'a'] = false;
                }

                if (!present[ch - 'a']) {
                    st.push(ch);
                    present[ch - 'a'] = true;
                }
            }

        }

        string res = "";
        while (!st.empty()) {
            res.push_back(st.top());
            st.pop();
        }

        reverse(res.begin(), res.end());
        return res;
        
    }
};

File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {
        int size = s.length();

        int[] lastOccurrence = new int[26];
        for (int i = 0; i < size; i++) {
            char ch = s.charAt(i);

            lastOccurrence[ch - 'a'] = i;
        }

        Stack<Character> st = new Stack<>();
        boolean[] isPresent = new boolean[26];
        for (int i = 0; i < size; i++) {
            char ch = s.charAt(i);

            if (isPresent[ch - 'a']) {

            } else {
                while (!st.isEmpty() && st.peek() > ch && lastOccurrence[st.peek() - 'a'] > i) {
                    isPresent[st.peek() - 'a'] = false;
                    st.pop();
                }

                st.push(ch);
                isPresent[ch - 'a'] = true;
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!st.isEmpty()) {
            char top = st.pop();

            sb.append(top);
        }
        
        return sb.reverse().toString();
    }
}